tests/test_basic_ops_dict.py

```diff
@@ -733,3 +735,23 @@ def test_setdefault() -> None:
     assert cfg["foo"] == 10
 
     assert cfg["foo"] == 10
+
+
+def test_set_valid_union_value() -> None:
+    cfg = OmegaConf.structured(DictUnion)
+    cfg.dict = {"a": 4, "b": 3.14}
+    assert isinstance(cfg.dict._get_node("a"), UnionNode)  # type: ignore
+    assert cfg.dict == {"a": 4, "b": 3.14}
+    assert isinstance(cfg.dict._get_node("a"), UnionNode)  # type: ignore
+    assert isinstance(cfg.dict._get_node("b"), UnionNode)  # type: ignore
+    assert cfg.dict.a == 4  # type: ignore
```

Comment:
I realized when looking at this line that although `cfg.dict` is declared to be of type `Dict`, OmegaConf actually turns it into a `DictConfig`.
@omry is this intended? (I don't think this has anything to do with this PR)